### PR TITLE
Initial connection pooling changes.

### PR DESCRIPTION
**Pull request by kylebrowning**

### CODE REVIEW

File: Sources/APNSKit/apns_kit.swift [hunk 1]
```diff
@@ -1,3 +1,29 @@
+import NIOAPNS
+import NIO
+import NIOKit
+
+public final class APNSKitConnectionManager: ConnectionPoolSource {
```
Comment (tanner0101) [inline Sources/APNSKit/apns_kit.swift]: I like the name `APNSConnectionSource` for this. That matches what other packages have done so far:

- `RedisConnectionSource`
- `PostgresConnectionSource`
- `MySQLConnectionSource`

File: Sources/APNSKit/apns_kit.swift [hunk 2]
```diff
@@ -1,3 +1,29 @@
...
+    public let eventLoop: EventLoop
+    private let config: APNSConfiguration
+    
+    deinit {
+        // Kill connection
```
Comment (tanner0101) [inline Sources/APNSKit/apns_kit.swift]: The connection pool will worry about closing connections when it is closed. No need to worry about that here.

File: Sources/APNSKit/apns_kit.swift [hunk 3]
```diff
@@ -1,3 +1,29 @@
...
+        self.eventLoop = eventLoop
+        self.config = config
+    }
+    public func makeConnection() -> EventLoopFuture<APNSConnection> {
+        return APNSConnection.connect(configuration: self.config, on: self.eventLoop)
+    }
+}
+extension APNSConnection: ConnectionPoolItem {
+    public var isClosed: Bool {
+        return false
```
Comment (tanner0101) [inline Sources/APNSKit/apns_kit.swift]: A todo / warning here would be useful.

File: Sources/APNSKit/apns_kit.swift [hunk 4]
```diff
@@ -1,3 +1,29 @@
+import NIOAPNS
```
Comment (tanner0101) [inline Sources/APNSKit/apns_kit.swift]: I think `NIOAPNS` and `NIOKit` should be `@_exported`. I also think that `NIOAPNS` should @_export anything that is part of it's public API (NIO, NIOSSL, etc). 

Doing `import NIOKit` (or `import NIOAPNS`), then having types that are public API appear undefined is confusing as an end user.

**Comment by kylebrowning** [comment]
Test pass locally. Looks to be some sort of open ssl issue again. :(

File: circle.yml [hunk 5]
```diff
@@ -3,15 +3,15 @@ version: 2
 jobs:
   linux:
     docker:
-      - image: codevapor/swift:5.0
+      - image: vapor/swift:5.0
```
Comment (tanner0101) [inline circle.yml]: This can be just `swift:5.0` now

Comment (kylebrowning) [reply circle.yml]: 👍

File: Sources/APNSKit/apns_kit.swift [hunk 6]
```diff
@@ -1,3 +1,23 @@
...
+}
+extension APNSConnection: ConnectionPoolItem {
+    public var isClosed: Bool {
+        // TODO: implement this.
+        return false
+    }
+}
+
+// Delete later when tests are in.
 struct apns_kit {
```
Comment (tanner0101) [inline Sources/APNSKit/apns_kit.swift]: We should remove this. Tests should be empty and passing for CI, right?

File: Sources/APNSKit/apns_kit.swift [hunk 7]
```diff
@@ -1,3 +1,23 @@
+public final class APNSConnectionSource: ConnectionPoolSource {
+    public let eventLoop: EventLoop
+    private let config: APNSConfiguration
```
Comment (tanner0101) [inline Sources/APNSKit/apns_kit.swift]: should be `configuration`.

File: Tests/APNSKitTests/APNSKitTests.swift [hunk 8]
```diff
@@ -3,10 +3,6 @@ import XCTest
 
 final class APNSKitTests: XCTestCase {
     func testExample() {
-        // This is an example of a functional test case.
-        // Use XCTAssert and related functions to verify your tests produce the correct
-        // results.
-        XCTAssertEqual(apns_kit().text, "Hello, World!")
     }
 
     static var allTests = [
```
Comment (tanner0101) [inline Tests/APNSKitTests/APNSKitTests.swift]: We should rm this and use `swift test --generate-linuxmain` instead.

Comment (kylebrowning) [reply Tests/APNSKitTests/APNSKitTests.swift]: @tanner0101 the whole file?

**Comment by kylebrowning** [comment]
```
$ swift test --generate-linuxmain
[2/2] Linking ./.build/x86_64-apple-macosx/debug/apns-kitPackageTests.xctest/Contents/MacOS/apns-kitPackageTests

kylebrowning at Kyles-iMac in ~/work/swift/apns-kit on connection-pooling [!]
$ swift test
Test Suite 'All tests' started at 2019-04-25 16:02:25.433
Test Suite 'All tests' passed at 2019-04-25 16:02:25.433.
	 Executed 0 tests, with 0 failures (0 unexpected) in 0.000 (0.000) seconds
```

File: Tests/APNSKitTests/APNSKitTests.swift (hunk 8 above)
Comment (tanner0101) [reply Tests/APNSKitTests/APNSKitTests.swift]: Just the `allTests` array. The command `swift test --generate-linuxmain` will use Swift / Objc magic to generate a `XCTestManifests.swift` for each test module and a `LinuxMain.swift` that combines them. 
Just make sure to add the newly generated files to version control after running the command.

Comment (kylebrowning) [reply Tests/APNSKitTests/APNSKitTests.swift]: OKay I think I did it right. 🤞

**Comment by penny-coin** [comment]
Hey @kylebrowning, you just merged a pull request, have a coin! 

You now have 4 coins.

**Pull request merged**